File: algorithmic_toolbox/week_3/02_greedy_algorithms_starter_files/fractional_knapsack/FractionalKnapsack.java
import java.util.Scanner;

public class FractionalKnapsack {

    private static int getMaxIndex(int[] weights, int[] values) {
        int max_i = 0;
        double max = 0;

        for (int i = 0; i < weights.length; i++) {
            if (weights[i] != 0 && (double) values[i] / weights[i] > max) {
                max = (double) values[i] / weights[i];
                max_i = i;
            }
        }
        return max_i;
    }

    private static double getOptimalValue(int capacity, int[] values, int[] weights) {
        double value = 0.0;

        for (int i = 0; i < weights.length; i++) {
            if (capacity == 0)
                return value;
            int index = getMaxIndex(weights, values);
            int a = Math.min(capacity, weights[index]);
            value += a * (double) values[index] / weights[index];
            weights[index] -= a;
            capacity -= a;
        }
        return value;
    }

    public static void main(String args[]) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int capacity = scanner.nextInt();
        int[] values = new int[n];
        int[] weights = new int[n];
        for (int i = 0; i < n; i++) {
            values[i] = scanner.nextInt();
            weights[i] = scanner.nextInt();
        }
        System.out.println(getOptimalValue(capacity, values, weights));
    }
}

File: algorithmic_toolbox/week_3/02_greedy_algorithms_starter_files/fractional_knapsack/fractional_knapsack.cpp
#include <iostream>
#include <vector>

using std::vector;

// finding the maximum index i.e the index for which values/weights is maximum
int get_max_index(vector<int> weights, vector<int> values) {
    int max_i = 0;
    double max = 0;

    for (int i = 0; i < weights.size(); i++) {
        if (weights[i] != 0 && (double) values[i] / weights[i] > max) {
            max = (double) values[i] / weights[i];
            max_i = i;  //maximum index
        }
    }
    return max_i;
}

// filling the knapsack staring from the maximum index
double get_optimal_value(int capacity, vector<int> weights, vector<int> values) {
    double value = 0.0;

    for (int i = 0; i < weights.size(); i++) {
        if (capacity == 0) return value; // capacity of the knapsack
        int index = get_max_index(weights, values);
        int a = std::min(capacity, weights[index]); 
        value += a * (double) values[index] / weights[index]; // total value of item placed in knapsack
        weights[index] -= a;
        capacity -= a;
    }

    return value;
}

int main() {
    int n;
    int capacity;
    std::cin >> n >> capacity;
    vector<int> values(n);
    vector<int> weights(n);
    for (int i = 0; i < n; i++) {
        std::cin >> values[i] >> weights[i];
    }

    double optimal_value = get_optimal_value(capacity, weights, values);

    std::cout.precision(10); //print the answer upto five decimal for error  corrections
    std::cout << optimal_value << std::endl;
    return 0;
}
